### PR TITLE
GHA refactor - refactor flows - [MOD-5837]

### PR DESCRIPTION
**Pull request by GuyAv46**
**Describe the changes in the pull request**

This PR defines the flows and events that trigger them for the repo, according to [the plan](https://redislabs.atlassian.net/wiki/spaces/DX/pages/3967844669/RediSearch+CI+refactor).
Not all the flows do what they are meant to, but the general structure of the CI is defined now, and all we are left to do is implement the remaining flows and define the branch protection rules (merge queue).

**Which issues this PR fixes**
1. MOD-5837

**Mark if applicable**

- [ ] This PR introduces API changes
- [ ] This PR introduces serialization changes

### CODE REVIEW

File: .github/workflows/flow-macos.yml
Comment (alonre24) [inline .github/workflows/flow-macos.yml:29]: nice

File: .github/workflows/common-flow-container.yml
Comment (alonre24) [inline .github/workflows/common-flow-container.yml:74]: Seems like this flow has a lot in common with common-flow-host... Consider making an even-more-common- flow.... Anyway, the same comments from common-flow-host applied here :)

File: .github/workflows/common-flow-host.yml
Comment (alonre24) [inline .github/workflows/common-flow-host.yml:22]: Are these required? when are we using `coordinator` or `standalone` as false?

File: .github/workflows/check-common-flow-sync.yml
Comment (alonre24) [inline .github/workflows/check-common-flow-sync.yml:15]: What is this flow?

**Review by alonre24 — COMMENTED** [review]
Great job, finally CI starting to look clear!!
1. Everything related to uploading artifacts to S3  will go in another PR, right? Note that we already have our AWS credentials set in github.
2. Please add to this PR a document describing the flows, including OSs and platforms that we are running on, the events that trigger the flows, the test variations running in each flow, and how to extract the tests' logs. The future team (including ourselves) will thank you for that...

File: .github/workflows/self-hosted.yml.TEMPLATE
Comment (GuyAv46) [reply .github/workflows/self-hosted.yml.TEMPLATE:12]: Any flow that we will want to run on an AWS instance

File: .github/workflows/event-pull_request.yml
Comment (GuyAv46) [reply .github/workflows/event-pull_request.yml:8]: Good question. I think this is a repository settings, but I will check

File: .github/workflows/common-flow-host.yml
Comment (GuyAv46) [reply .github/workflows/common-flow-host.yml:24]: There are some differences, this values are not suppose to be inputs so it better to define them here. What do you mean by “where it is required”?

File: .github/workflows/common-flow-host.yml
Comment (GuyAv46) [reply .github/workflows/common-flow-host.yml:82]: We can. At first I wasn’t sure what will be the conditions for continuing, and I think it will be easier to change the conditions this way, but for now it will be enought

File: .github/workflows/common-flow-host.yml
Comment (GuyAv46) [reply .github/workflows/common-flow-host.yml:124]: Because currently there are some logs generated. I made another PR that fixes what causes these logs, until it is merged we don’t want to fail the flow (but we probably will in the future)

File: .github/workflows/event-pull_request.yml
Comment (GuyAv46) [reply .github/workflows/event-pull_request.yml:4]: For now yes, because we can skip it based on what files where edited, making PRs safer (so you can’t accidentally push untested code)

File: .github/workflows/event-merge_queue.yml
Comment (GuyAv46) [reply .github/workflows/event-merge_queue.yml:8]: :(

File: .github/workflows/merge-nightly-common.yml
Comment (GuyAv46) [reply .github/workflows/merge-nightly-common.yml:25]: Currently it’s for coordinator only, but it can have any special options here

File: .github/workflows/event-nightly.yml
Comment (GuyAv46) [reply .github/workflows/event-nightly.yml:25]: Right. It should be implemented in the next few weeks

File: .github/workflows/event-nightly.yml
Comment (GuyAv46) [reply .github/workflows/event-nightly.yml:22]: Agree. Let’s see

File: .github/workflows/event-merge_queue.yml
Comment (GuyAv46) [reply .github/workflows/event-merge_queue.yml:5]: As far as we understood, this flow defines what checks should pass for a PR to be merged. This will replace defining some flows as “required” before pressing “merge”, and instead of labeling (which is kind of a manual way to do the same thing)

File: .github/workflows/event-weekly.yml
Comment (GuyAv46) [reply .github/workflows/event-weekly.yml:25]: Right, both are planned for a later PR

File: .github/workflows/flow-linux-platforms.yml
Comment (GuyAv46) [reply .github/workflows/flow-linux-platforms.yml:5]: This list was used in the original flow. Good catch!

File: .github/workflows/flow-self-hosted-arm.yml
Comment (GuyAv46) [reply .github/workflows/flow-self-hosted-arm.yml:21]: One can dream..  they might already support it but for enterprise only

File: .github/workflows/event-pull_request.yml
Comment (GuyAv46) [reply .github/workflows/event-pull_request.yml:8]: found the setting [here](https://docs.github.com/en/repositories/managing-your-repositorys-settings-and-features/enabling-features-for-your-repository/managing-github-actions-settings-for-a-repository#enabling-workflows-for-forks-of-private-repositories). seems to be diable by default

File: .github/workflows/common-flow-host.yml
Comment (GuyAv46) [reply .github/workflows/common-flow-host.yml:22]: Not required and default to `true`, but if one is false you can test only one of the configurations (the other one).
It is used now for running specific configurations that are only relevant to coordinator.

File: .github/workflows/common-flow-host.yml
Comment (GuyAv46) [reply .github/workflows/common-flow-host.yml:124]: [This PR](https://github.com/RediSearch/RediSearch/pull/3784) will fix the warning, and then we could uncomment this verification

File: .github/workflows/common-flow-container.yml
Comment (GuyAv46) [reply .github/workflows/common-flow-container.yml:74]: sadly it is immpossible, because GHA does not allow refering to `container:` without supplying a container, so I had to duplicate this file, with specific differences in the beggining of the file.
As an alternative (until GHA will support not refering to a container and behaving as running directly on the runner), I made `check-common-flow-sync.yml` flow, that verify that the logic part of both files are in sync, making it immposible to edit one and not theother, without failing the CI.

File: .github/workflows/check-common-flow-sync.yml
Comment (GuyAv46) [reply .github/workflows/check-common-flow-sync.yml:15]: Because GHA does not allow referring to `container:` without supplying a container, I had to duplicate this file, with specific differences at the beginning of the file.
As an alternative (until GHA will support not referring to a container and behaving as running directly on the runner), I made check-common-flow-sync.yml flow, which verifies that the logic part of both files is in sync, making it impossible to edit one and not the other, without failing the CI.

File: .github/workflows/merge-nightly-common.yml
Comment (GuyAv46) [reply .github/workflows/merge-nightly-common.yml]: I did but did not verify. We had to add:
```yml
env:
  REDISEARCH_MT_BUILD: ${{ vars.REDISEARCH_MT_BUILD }}
```
fixed

File: .github/workflows/self-hosted.yml.TEMPLATE
Comment (alonre24) [reply .github/workflows/self-hosted.yml.TEMPLATE:12]: PLease add this line (the purpose of this template) to the top of the file

File: .github/workflows/event-pull_request.yml
Comment (alonre24) [reply .github/workflows/event-pull_request.yml:8]: I see. I would change it to the more restricted policy (required approval to run workflow **every time** for external users). WDYT?

File: .github/workflows/common-flow-host.yml
Comment (alonre24) [reply .github/workflows/common-flow-host.yml:24]: Where does the `sudo` required?

File: .github/workflows/common-flow-host.yml
Comment (alonre24) [reply .github/workflows/common-flow-host.yml:125]: I see, so let's merge it and uncomment

File: .github/workflows/common-flow-host.yml
Comment (GuyAv46) [reply .github/workflows/common-flow-host.yml:24]: oh. it is probably not used right now, but some environments require `sudo` while others don't recognize the command so it will be needed when we remove readies

**Review by alonre24 — APPROVED** [review]

**Pull request merged**
